tutorials/cloud-iot-rtdp/streaming/src/main/java/com/google/cloud/solutions/rtdp/Simulator.java

```diff
@@ -0,0 +1,162 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * 
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.cloud.solutions.rtdp;
+
+import io.jsonwebtoken.JwtBuilder;
+import io.jsonwebtoken.Jwts;
+import io.jsonwebtoken.SignatureAlgorithm;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.security.KeyFactory;
+import java.security.PrivateKey;
+import java.security.spec.PKCS8EncodedKeySpec;
+import java.util.Random;
+import org.eclipse.paho.client.mqttv3.IMqttDeliveryToken;
+import org.eclipse.paho.client.mqttv3.MqttCallback;
+import org.eclipse.paho.client.mqttv3.MqttClient;
+import org.eclipse.paho.client.mqttv3.MqttConnectOptions;
+import org.eclipse.paho.client.mqttv3.MqttException;
+import org.eclipse.paho.client.mqttv3.MqttMessage;
+import org.eclipse.paho.client.mqttv3.MqttPersistenceException;
+import org.eclipse.paho.client.mqttv3.persist.MemoryPersistence;
+import org.joda.time.DateTime;
+
+/*
+ * Simulator emulates multiple MQTT clients and establishes sessions to Cloud IoT Core.
+ * It randomly chooses MQTT sessions and sends a specified number of messages repeatedly.
+ */
+public class Simulator implements MqttCallback {
+  public static final int NUM_DEVICES = 30;
+
+  private static PrivateKey loadKeyFile(String filename, String algorithm) throws Exception {
+    byte[] keyBytes = Files.readAllBytes(Paths.get(filename));
+    PKCS8EncodedKeySpec spec = new PKCS8EncodedKeySpec(keyBytes);
+    KeyFactory kf = KeyFactory.getInstance(algorithm);
+    return kf.generatePrivate(spec);
+  }
+
+  private static String createJwt(String projectId, String privateKeyFile, String algorithm)
+      throws Exception {
+    DateTime now = new DateTime();
+    JwtBuilder jwtBuilder =
+        Jwts.builder()
+            .setIssuedAt(now.toDate())
+            .setExpiration(now.plusMinutes(60).toDate())
+            .setAudience(projectId);
+
+    if (algorithm.equals("RS256")) {
+      PrivateKey privateKey = loadKeyFile(privateKeyFile, "RSA");
+      return jwtBuilder.signWith(SignatureAlgorithm.RS256, privateKey).compact();
+    } else if (algorithm.equals("ES256")) {
+      PrivateKey privateKey = loadKeyFile(privateKeyFile, "EC");
+      return jwtBuilder.signWith(SignatureAlgorithm.ES256, privateKey).compact();
+    } else {
+      throw new IllegalArgumentException(
+          "Invalid algorithm " + algorithm + ". Should be one of 'RS256' or 'ES256'.");
+    }
+  }
+
+  private MqttClient client;
+  private String deviceId;
+  private String topic;
+  private Random r;
+  private String sz = "20";
```

Comment:
Lint is telling me to make these final, optionally set them to final if they won't change.